README.md
# simplifa11y

> a11y-ready react component library using styled-components for theming

trying to make this as barebones as possible while keeping decent customization

## [demo](https://pipedr3am.github.io/simplifa11y)

## install

```bash
yarn add @pipedr3am/simplifa11y -s
```

## License

GNU General Public License v3.0 © [dr-mimic](https://github.com/dr-mimic)

### code adapted/refactored from
- https://blog.logrocket.com/how-to-build-an-accessible-date-picker-component-in-react/
- https://github.com/jkithome/accessible-datepicker


CHANGELOG.md
## [0.0.7] - 2021-10-23
### Fixed
- next and previous year controls now set date && month for min max values

### Removed
- storybook removed in favor of barebones examples with `react-code-blocks` for code examples

## [0.0.7-rc.0] - 2021-10-21
### Added
- parameterized aria-label for formatHint as `ariaLabelFormatHint`
- if cur date > max date, cal ui will start on max date not current
- if cur date < max date, cal ui will start on cur date
- on entering a full date will submit (enter key) via date input and clear errors

### Fixed
- updating initialDate after an API call now triggers correctly
- fixed bug with having multiple datepickers on pages
- backspace fixed when going back from added slashes i.e. `02/02/` <-- will now give `02/0`

### Changed
- months for aria-labels are now full i.e. January vs Jan

## [0.0.6-rc.10] - 2021-10-06
### Lost
- rc.10 and rc.11 updates were lost to the void...

## [0.0.6-rc.9] - 2021-10-05
### Fixed
- enter logic on DateInput fixes, removed the spacing around the '/'

## [0.0.6-rc.8] - 2021-09-29
### Added
- inputId param added to override default date input text id
- on receiving a full date, input box now fires set dateString method up component
  - calls `onInvalidDate` error up if invalid dateRange

### Fixed
- calendar td button types set to 'button' to prevent form submitions via formik
- next previous year month button types set to 'button' to prevent form submitions via formik
- calendar icon button type set to 'button' to prevent form submitions via formik
- removed year / month arrow button backgrounds and borders
- input <enter> will skip format and validation check if character length is not 14, will instead call the error callback
- invalid date input check will skip format if character length is not 14
- valid date checks on every keystroke fire errors up to parent
- hitting enter after only entering partial date input will no longer throw error

### Changed
- added better support for screen-readers, structure updates and tabbing control
  - added aria-label 'toggle calendar' to calendar icon 
  - calendar icon changed from li to div to optimize what screen reader announces
  - added aria-describedby for date input pointing to inputHint
  - added aria-label for inputHint as `format is ${inputHint}`

### Security
- security updates

## [0.0.5] - 2021-09-20
### Added
- onInvalidDate callback sends out {message, details: { dateRange, dateInput }}
- date validation on input enter (date range check)

### Changed
- dateRange no longer required, when left out year limits are 0000 and 9999

## [0.0.5-rc.3] - 2021-09-20
### Added
- outline param added for global wrap
- outlineHover param added for global wrap
- dateRange min and max constrains left & right month & year functionality

## [0.0.5-rc.2] - 2021-09-20
### Fixed
- dateRange year working and selects max year passed in

## [0.0.5-rc.1] - 2021-09-19
### Added
- zIndex added for calendar to solve absolute positioning issues on parent apps
- width added for calendar
- dateRange added for minimum and maximum dates
- custom theme mods available for single parameter updates to defaultTheme
- theme parameter 'none' available for using plain css 
- padding param added to dateInput and inputHint
- width param added to dateInput

## [0.0.5-rc.0] - 2021-09-15
### Fixed
- backgroundColor changed to background for Calendar solid bg

## [0.0.4] - 2021-09-15
### Added
- backgroundColor param added to defaultTheme

## [0.0.3] - 2021-09-15
### Removed
- removed @axe-core from packages

## [0.0.1] - 2021-09-15
### Added
- theming added for core date text and backgrounds

### Changed
- removed all css in favor of styled-components

### Removed
- lodash removed

### Fixed
- selecting date no longer selects previous date (local timezone / format fix)

src/components/Calendar.js
import React, { useState, useEffect } from 'react'
import {
  format as fnsFormat,
  startOfMonth,
  getDaysInMonth,
  getWeeksInMonth,
  getDay,
  getMonth,
  getYear,
  setDate,
  getDate,
  isEqual,
  toDate
} from 'date-fns'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faAngleLeft, faAngleRight, faAngleDoubleLeft, faAngleDoubleRight } from '@fortawesome/free-solid-svg-icons'

import * as S from './Calendar.styles'
import { dateFromString } from '../helpers/formatDate'
import makeKeyControl from '../helpers/calendar/keyControl'

const ariaDays = [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday'
]

const ariaMonths = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December'
]

const Calendar = ({
  date,
  format,
  handleSelectDate,
  closeCalendar,
  dateRange,
  calendarShown
}) => {
  const [selectedDate, setSelectedDate] = useState(dateFromString({ date }))
  const [focusDay, setFocusDay] = useState({
    focus: true,
    previousDate: ''
  })
  const keyControl = makeKeyControl({
    format,
    dateRange,
    selectedDate,
    setSelectedDate,
    setFocusDay
  })

  useEffect(() => {
    if (calendarShown && focusDay.focus) {
      setFocusDay({ focus: false })
      keyControl.focusSelectedDate({ previousDate: focusDay.previousDate })
    }
  }, [calendarShown, focusDay])

  const previousYearKeyPress = e => {
    const keyCode = e.keyCode
    const control = e.shiftKey

    if (keyCode === 13 || keyCode === 32) {
      e.preventDefault()
      keyControl.setPreviousYear()
    } else if (keyCode === 9) { // TAB
      if (control) {
        // shift held
        e.preventDefault()
        if (document.activeElement === document.getElementById('previousYear')) {
          keyControl.focusSelectedDate({ previousDate: '' })
        }
      }
    }
  }

  // @TODO move to keyControl + unit tests
  const handleDayKeyPress = e => {
    e.preventDefault()
    const keyCode = e.keyCode
    const control = e.shiftKey
    switch (keyCode) {
      case 9: // TAB
        if (!control) {
          document.getElementById('previousYear').focus()
          e.stopPropagation()
          e.preventDefault()
        }
        break
      case 13: // Enter
        handleSelectDate(fnsFormat(selectedDate, format))
        return
      case 27: // Esc
        closeCalendar()
        return
      case 32: // Space
        handleSelectDate(fnsFormat(selectedDate, format))
        return
      case 33: // Page Up
        control ? keyControl.setDatePreviousYear() : keyControl.setDatePreviousMonth()
        return
      case 34: // Page Down
        control ? keyControl.setDateNextYear() : keyControl.setDateNextMonth()
        return
      case 35: // End
        keyControl.setMonthEnd()
        return
      case 36: // Home
        keyControl.setMonthStart()
        return
      case 37: // Left
        keyControl.setPreviousDay()
        return
      case 38: // Up
        keyControl.setPreviousWeek()
        return
      case 39: // Right
        keyControl.setNextDay()
        return
      case 40: // Down
        keyControl.setNextWeek()
        break
      default:
        break
    }
  }
  const handleDateSelection = (date) => {
    const dateString = fnsFormat(date, format)
    handleSelectDate(dateString)
  }

  // @TODO refactor this, extract + unit test
  const makeWeek = ({ selectedDate, weekNum, daysInMonth, startWeekday }) => {
    const curYear = toDate(new Date(selectedDate)).getFullYear()
    const curMonth = toDate(new Date(selectedDate)).getMonth() + 1
    let minDay = 0
    let maxDay = 0
    let minMonth = 0
    let maxMonth = 0
    let minYear = 0
    let maxYear = 0

    const [miMonth, miDay, miYear] = dateRange ? dateRange.min.split('/') : ['01', '01', '0000']
    const [maMonth, maDay, maYear] = dateRange ? dateRange.max.split('/') : ['01', '01', '9999']
    minDay = parseInt(miDay)
    maxDay = parseInt(maDay) + 1
    minMonth = parseInt(miMonth)
    maxMonth = parseInt(maMonth)
    minYear = parseInt(miYear)
    maxYear = parseInt(maYear)

    const inMinYear = curYear === minYear
    const inMaxYear = curYear === maxYear
    const yearValid = curYear >= minYear && curYear <= maxYear

    const returnArr = []
    let dayNum = weekNum === 0 ? 1 : (weekNum * 7) - startWeekday + 1
    for (let i = 0; i < 7; i++) {
      let minDayValid = true
      let maxDayValid = true
      const inMinMonth = curMonth === minMonth
      const inMaxMonth = curMonth === maxMonth

      if (inMinMonth && yearValid) {
        minDayValid = dayNum > minDay
      }
      if (inMaxMonth && yearValid) {
        maxDayValid = dayNum < maxDay
      }
      let validDay = true

      if (curMonth < minMonth && inMinYear) {
        validDay = false
      } else if (curMonth > maxMonth && inMaxYear) {
        validDay = false
      }

      if (curYear < minYear || curYear > maxYear) {
        validDay = false
      }

      if (weekNum === 0) {
        if (i < startWeekday) {
          returnArr.push(null)
          continue
        }
      }
      if (dayNum <= daysInMonth) {
        if (validDay) {
          returnArr.push(setDate(selectedDate, dayNum))
          dayNum++
        } else {
          returnArr.push(null)
          dayNum++
        }
      } else {
        returnArr.push(null)
      }
    }
    return returnArr
  }

  const getWrittenDay = ({ day }) => {
    const dateString = fnsFormat(day, format)
    const dateObj = new Date(dateString)
    const dayIdx = getDay(dateObj)
    const monthIdx = getMonth(dateObj)
    const year = getYear(dateObj)

    const writtenDay = ariaDays[dayIdx]
    const writtenMonth = ariaMonths[monthIdx]
    return `${writtenDay} ${writtenMonth} ${year}`
  }

  // @TODO same refactor somewhere
  const generateMonth = () => {
    const daysInMonth = getDaysInMonth(selectedDate)
    const weeksInMonth = getWeeksInMonth(selectedDate)
    const startWeekday = getDay(startOfMonth(selectedDate))

    const weeks = []
    for (let i = 0; i < weeksInMonth; i++) {
      weeks.push(makeWeek({ selectedDate, weekNum: i, daysInMonth, startWeekday }))
    }

    return weeks
  }

  const onBlur = e => {
    // @Recall was trying to close calendar on clicking off of it ... getting there
    // leaving this as its not needed atm
    // console.log('blur')
    // if user is not focusing on a calendar button, they have clicked outside the calendar
    // const activeFocus = document.activeElement['data-focus']
    // console.log('activeFocus:', document.activeElement)
    // if (!activeFocus) {
    //   console.log('hide ui')
    // }

  }

  return (
    <S.Calendar
      id='calendarDisplay'
      aria-expanded={calendarShown}
      onBlur={onBlur}
    >
      <S.Title>
        <S.Icons>
          <S.IconWrap
            type='button'
            id='previousYear'
            data-focus='cal'
            aria-label='previous year'
            onClick={keyControl.setPreviousYear}
            onKeyDown={(e) => previousYearKeyPress(e)}
          >
            <FontAwesomeIcon icon={faAngleDoubleLeft} />
          </S.IconWrap>
          <S.IconWrap
            type='button'
            data-focus='cal'
            aria-label='previous month'
            onClick={keyControl.setPreviousMonth}
          >
            <FontAwesomeIcon icon={faAngleLeft} />
          </S.IconWrap>
        </S.Icons>
        <S.Month role='heading'>
          <b>
            {fnsFormat(selectedDate, 'MMMM yyyy')}
          </b>
        </S.Month>
        <S.Icons>
          <S.IconWrap
            type='button'
            data-focus='cal'
            aria-label='next month'
            onClick={keyControl.setNextMonth}
            onKeyUp={(e) => keyControl.handleKeyPress({ e, callback: keyControl.setNextMonth })}
          >
            <FontAwesomeIcon icon={faAngleRight} />
          </S.IconWrap>
          <S.IconWrap
            type='button'
            data-focus='cal'
            aria-label='next year'
            onClick={keyControl.setNextYear}
            onKeyUp={(e) => keyControl.handleKeyPress({ e, callback: keyControl.setNextYear })}
          >
            <FontAwesomeIcon icon={faAngleDoubleRight} />
          </S.IconWrap>
        </S.Icons>
      </S.Title>
      <table
        role='grid'
        aria-label='select day via arrow keys'
      >
        <thead>
          <tr role='row'>
            <S.DaysOfWeek role='columnheader' aria-label='Sunday'><abbr title='Sunday'>Su</abbr></S.DaysOfWeek>
            <S.DaysOfWeek role='columnheader' aria-label='Monday'><abbr title='Monday'>Mo</abbr></S.DaysOfWeek>
            <S.DaysOfWeek role='columnheader' aria-label='Tuesday'><abbr title='Tuesday'>Tu</abbr></S.DaysOfWeek>
            <S.DaysOfWeek role='columnheader' aria-label='Wednesday'><abbr title='Wednesday'>We</abbr></S.DaysOfWeek>
            <S.DaysOfWeek role='columnheader' aria-label='Thursday'><abbr title='Thursday'>Th</abbr></S.DaysOfWeek>
            <S.DaysOfWeek role='columnheader' aria-label='Friday'><abbr title='Friday'>Fr</abbr></S.DaysOfWeek>
            <S.DaysOfWeek role='columnheader' aria-label='Saturday'><abbr title='Saturday'>Sa</abbr></S.DaysOfWeek>
          </tr>
        </thead>
        <tbody>
          {generateMonth().map((week, i) => (
            <tr className='week' key={`week-${i}`} role='row'>
              {week.map((day, i) => (day
                ? <S.Cell
                    id={`${getDate(day)} ${getWrittenDay({ day })}`}
                    aria-selected={isEqual(selectedDate, day)}
                    $selected={isEqual(selectedDate, day)}
                    key={`day-cell-${i}`}
                    onClick={() => handleDateSelection(day)}
                  >
                  <button
                    type='button'
                    id={fnsFormat(day, format)}
                    data-focus='cal'
                    aria-label={`${getDate(day)} ${getWrittenDay({ day })}`}
                    onKeyDown={handleDayKeyPress}
                    tabIndex={`${isEqual(selectedDate, day) ? '0' : '-1'}`}
                  >
                    {getDate(day)}
                  </button>
                </S.Cell>
                : <S.Cell $empty key={`day-cell-${i}`}>&nbsp;</S.Cell>
              ))}
            </tr>
          ))}
        </tbody>
      </table>
    </S.Calendar>
  )
}
export default Calendar
